onadata/libs/utils/middleware.py

```diff
@@ -64,3 +75,20 @@ def __call__(self, request):
                     query['time'], " ".join(query['sql'].split())))
 
         return response
+
+
+class OperationalErrorMiddleware(BaseMiddleware):
+    """
+    Captures requests returning 500 status code.
+    Then retry it against master database.
+    """
+    def process_exception(self, request, exception):
+        # Filter out OperationalError Exceptions
+        if isinstance(exception, OperationalError):
+            already_raised = getattr(settings, "ALREADY_RAISED", False)
```

Comment:
I think we need to limit the middleware to very specific `OperationaError` exceptions. Specifically the ones that contain "canceling statement due to conflict with recovery".